bcm47xx: add support for Netgear R6200 V1

motolav opened the pull request:
The Netgear R6200 V1 is a 2.4/5 GHz band 11ac router, based on the Broadcom BCM4718A1.

Specification:

CPU: BCM4718A1 480 MHz. MIPS 74K
RAM: 128 MB
Flash: 16 MB
Ethernet: 5 Ports all 10/100/1000 Mbps, 1 WAN + 4 LAN
WiFi: BCM4718A1(2.4GHz) + BCM4352(5GHz)
USB: 1 USB 2.0
LEDs: Power, Internet, Wireless, USB and a Logo LED
Switches: Reset, Wireless, WPS
Other: 1 UART, 1 I2C

Product Specification:
https://wikidevi.com/wiki/Netgear_R6200v1
The Offical Netgear page only covers the V2 Specifications

Installation:
1. Boot R6200V1 Normally
2. Access the Router's Admin page should be "http://192.168.1.1/"
3. Go to the "Advanced" tab on the top and inside "Administration" on the side bar is "Router Update"
4. Select the OpenWrt build with the name netgear-r6200-v1-squashfs.chk and upload to device
5. Wait for firmware to finish flashing  

Issues:
5Ghz Band is unsupported due to it needing a Propietary driver (BCM4360)
LEDs for the Logo LED, Power Status, Wireless Status, and USB status are Controlled by a 74x164 shift register that the original firmware controlled via bit banging.
As of now the power LED stays amber and the WAN indicator is controlled by the ethernet switch driver.

Notes:
My device has the model number U12H19T01 on the board but the factory firmware updater recognizes only U12H19T00, which matches the model in the original firmware, with the region set to 1 as the proper firmware for a Netgear R6200 V1 from North America.

Signed-off-by: Edward Matijevic motolav@gmail.com

motolav commented:
I noticed I'm not fully compliant with the rules for submitting patches and am fixing my commits

motolav commented:
I have added the changes to the 4.19 patch for brcm47xx and have tested that it boots as well so I'm done with my changes as I can't get the LEDs to work properly currently.
I'm done with the PR and sorry its a bit of a mess, Its my first ever PR.

hauke commented:
The code looks ok to me.

Could you please squash this together into one patch.
Please run "make target/linux/{clean,refresh} V=99" to refresh the patches after your changes.

Please also send the support for your new device to the upstream Linux kernel so we do not have to maintain it here.

mwarning commented on target/linux/brcm47xx/image/Makefile:
```diff
@@ -822,6 +822,15 @@ define Device/netgear-wgr614-v10-na
 TARGET_DEVICES += netgear-wgr614-v10-na
 
+define Device/netgear-r6200-v1
+  DEVICE_TITLE := Netgear R6200 V1
```
Please split up DEVICE_TITLE:
```
DEVICE_VENDOR := Netgear
DEVICE_MODEL := R6200
DEVICE_VARIANT := V1
```
See https://github.com/openwrt/openwrt/pull/2124 - thanks!

mwarning replied:
ah, I forgot that it is NETGEAR (upper case). The other tickets are already doing it. e.g. https://github.com/openwrt/openwrt/pull/2222.

motolav replied:
Thanks, I fixed it in my latest push. I didn't notice it in the top of that thread earlier

motolav commented:
I have updated the makefile and split the patch for R6200 support.

adrianschmutzler commented:
@hauke Maybe have another look at this ...

motolav commented:
My kernel patches were added into mips-next and are waiting for the 5.4 merge window.
Here is the serial console output from my test device:
```
Decompressing..........done
Decompressing..........done


CFE for R6200 version: v1.0.1
Build Date: Fri May 11 13:46:43 CST 2012 
Init Arena
Init Devs.
Boot partition size = 262144(0x40000)
Found an ST compatible serial flash with 256 64KB blocks; total size 16MB
et0: Broadcom BCM47XX 10/100/1000 Mbps Ethernet Controller 2010.09.30.0
CPU type 0x19740: 480MHz
Tot mem: 131072 KBytes

Device eth0:  hwaddr 84-1B-5E-F5-37-41, ipaddr 192.168.1.1, mask 255.255.255.0
        gateway not set, nameserver not set
Checksum length overflow!
Loader:raw Filesys:raw Dev:flash0.os File: Options:(null)
Loading: .. 3832 bytes read
Entry at 0x80001000
Closing network.
Starting program at 0x80001000
[    0.000000] Linux version 4.14.109 (builduser@buildhost) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r9736-04ac8b0151)) #0 Wed Apr 3 04:27:22 2019
[    0.000000] CPU0 revision is: 00019740 (MIPS 74Kc)
[    0.000000] bcm47xx: Using bcma bus
[    0.000000] (NULL device *): bus0: Found chip with id 0x4716, rev 0x01 and package 0x0A
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 07fff000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] This processor doesn't support highmem. -131068k highmem ignored
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffefff]
[    0.000000]   HighMem  empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffefff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffefff]
[    0.000000] random: get_random_bytes called from start_kernel+0x90/0x478 with crng_init=0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32511
[    0.000000] Kernel command line: noinitrd console=ttyS0,115200
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 124816K/131068K available (3587K kernel code, 147K rwdata, 808K rodata, 184K init, 293K bss, 6252K reserved, 0K cma-reserved, 0K highmem)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 128
[    0.000000] (NULL device *): bus0: Core 0 found: ChipCommon (manuf 0x4BF, id 0x800, rev 0x1F, class 0x0)
[    0.000000] (NULL device *): bus0: Core 1 found: IEEE 802.11 (manuf 0x4BF, id 0x812, rev 0x11, class 0x0)
[    0.000000] (NULL device *): bus0: Core 2 found: GBit MAC (manuf 0x4BF, id 0x82D, rev 0x00, class 0x0)
[    0.000000] (NULL device *): bus0: Core 3 found: MIPS 74K (manuf 0x4A7, id 0x82C, rev 0x01, class 0x0)
[    0.000000] (NULL device *): bus0: Core 4 found: USB 2.0 Host (manuf 0x4BF, id 0x819, rev 0x04, class 0x0)
[    0.000000] (NULL device *): bus0: Core 5 found: PCIe (manuf 0x4BF, id 0x820, rev 0x0E, class 0x0)
[    0.000000] (NULL device *): bus0: Core 6 found: DDR1/DDR2 Memory Controller (manuf 0x4BF, id 0x82E, rev 0x01, class 0x0)
[    0.000000] (NULL device *): bus0: Core 7 found: Internal Memory (manuf 0x4BF, id 0x80E, rev 0x07, class 0x0)
[    0.000000] (NULL device *): bus0: Core 8 found: I2S (manuf 0x4BF, id 0x834, rev 0x00, class 0x0)
[    0.000000] (NULL device *): bus0: Found M25FL128 serial flash (size: 16384KiB, blocksize: 0x10000, blocks: 256)
[    0.000000] (NULL device *): bus0: Early bus registered
[    0.000000] MIPS: machine is Netgear R6200 V1
[    0.000000] bcm47xx: Setting up vectored interrupts
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7963585194 ns
[    0.000016] sched_clock: 32 bits at 240MHz, resolution 4ns, wraps every 8947848189ns
[    0.000094] Calibrating delay loop... 239.61 BogoMIPS (lpj=1198080)
[    0.070027] pid_max: default: 32768 minimum: 301
[    0.070408] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.070444] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.073890] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.073940] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.074735] NET: Registered protocol family 16
[    0.107046] clocksource: Switched to clocksource MIPS
[    0.108895] NET: Registered protocol family 2
[    0.110151] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.110285] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.110328] TCP: Hash tables configured (established 1024 bind 1024)
[    0.110563] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.110615] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.111002] NET: Registered protocol family 1
[    0.528188]  2(S)*
[    0.528195]  3 
[    0.528216]  4 
[    0.528230]  5 
[    0.528244]  6 
[    0.528257]  D 
[    0.528271]  I 
[    0.528283] 
[    0.528312]  2(S) 
[    0.528318]  3*
[    0.528332]  4 
[    0.528345]  5 
[    0.528358]  6 
[    0.528371]  D 
[    0.528384]  I 
[    0.528396] 
[    0.528424]  2(S) 
[    0.528431]  3 
[    0.528444]  4*
[    0.528457]  5 
[    0.528470]  6 
[    0.528483]  D 
[    0.528496]  I 
[    0.528508] 
[    0.528535]  2(S) 
[    0.528542]  3 
[    0.528555]  4 
[    0.528568]  5 
[    0.528581]  6 
[    0.528594]  D 
[    0.528607]  I*
[    0.528619] 
[    0.528647]  2(S) 
[    0.528653]  3 
[    0.528667]  4 
[    0.528680]  5*
[    0.528693]  6 
[    0.528706]  D 
[    0.528718]  I 
[    0.528731] 
[    0.528759]  2(S) 
[    0.528765]  3 
[    0.528779]  4 
[    0.528792]  5 
[    0.528804]  6*
[    0.528817]  D 
[    0.528830]  I 
[    0.528842] 
[    0.528871]  2(S) 
[    0.528877]  3 
[    0.528890]  4 
[    0.528903]  5 
[    0.528916]  6 
[    0.528929]  D*
[    0.528942]  I 
[    0.528954] 
[    0.528981]  2(S) 
[    0.528987]  3 
[    0.529001]  4 
[    0.529014]  5 
[    0.529027]  6 
[    0.529039]  D 
[    0.529052]  I*
[    0.529064] 
[    0.529092]  2(S)*
[    0.529098]  3 
[    0.529111]  4 
[    0.529124]  5 
[    0.529137]  6 
[    0.529150]  D 
[    0.529163]  I 
[    0.529175] 
[    0.529201] (NULL device *): bus0: PCIEcore in host mode found
[    0.732340] PCI host bridge to bus 0000:00
[    0.732392] pci_bus 0000:00: root bus resource [mem 0x08000000-0x0bffffff]
[    0.732420] pci_bus 0000:00: root bus resource [io  0x0100-0x07ff]
[    0.732447] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.732475] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.732584] bcma: PCI: Fixing up bridge 0000:00:00.0
[    0.732607] bcma: PCI: Fixing up device 0000:00:00.0
[    0.732736] pci 0000:00:00.0: enabling Extended Tags
[    0.733264] bcma: PCI: Fixing up bridge 0000:00:00.1
[    0.733291] bcma: PCI: Fixing up device 0000:00:00.1
[    0.733421] pci 0000:00:00.1: enabling Extended Tags
[    0.734124] bcma: PCI: Fixing up addresses 0000:00:01.0
[    0.734175] pci 0000:00:01.0: BAR 0: assigned [mem 0x08000000-0x08007fff 64bit]
[    0.736349] (NULL device *): bus0: Bus registered
[    0.737623] Crashlog allocated RAM at address 0x3f00000
[    0.744181] workingset: timestamp_bits=30 max_order=15 bucket_order=0
[    0.752510] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.752543] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.769830] io scheduler noop registered
[    0.769866] io scheduler deadline registered (default)
[    0.770293] Serial: 8250/16550 driver, 2 ports, IRQ sharing enabled
[    0.771206] console [ttyS0] disabled
[    0.791452] serial8250.0: ttyS0 at MMIO 0xb8000300 (irq = 2, base_baud = 1250000) is a U6_16550A
[    1.456598] console [ttyS0] enabled
[    1.654535] 12 bcm47xxpart partitions found on MTD device bcm47xxsflash
[    1.661327] Creating 12 MTD partitions on "bcm47xxsflash":
[    1.666915] 0x000000000000-0x000000040000 : "boot"
[    1.676935] 0x000000040000-0x000000f40000 : "firmware"
[    1.683629] 3 trx partitions found on MTD device firmware
[    1.689219] Creating 3 MTD partitions on "firmware":
[    1.694299] 0x00000000001c-0x000000000948 : "loader"
[    1.703493] 0x000000000948-0x00000017a400 : "linux"
[    1.710770] 0x00000017a400-0x000000f00000 : "rootfs"
[    1.717270] mtd: device 4 (rootfs) set to be root filesystem
[    1.723091] 1 squashfs-split partitions found on MTD device rootfs
[    1.729460] 0x0000005e0000-0x000000f00000 : "rootfs_data"
[    1.739929] 0x000000f40000-0x000000f50000 : "ML"
[    1.746845] 0x000000f50000-0x000000f60000 : "ML"
[    1.755469] 0x000000f60000-0x000000f70000 : "ML"
[    1.762428] 0x000000f70000-0x000000f80000 : "ML"
[    1.771029] 0x000000f80000-0x000000f90000 : "ML"
[    1.778416] 0x000000f90000-0x000000fa0000 : "ML"
[    1.786523] 0x000000fa0000-0x000000fd0000 : "ML"
[    1.793531] 0x000000fd0000-0x000000fe0000 : "POT"
[    1.802209] 0x000000fe0000-0x000000ff0000 : "board_data"
[    1.809902] 0x000000ff0000-0x000001000000 : "nvram"
[    1.821678] libphy: Fixed MDIO Bus: probed
[    1.826040] bgmac_bcma bcma0:2: Found PHY addr: 30 (NOREGS)
[    1.914704] b53_common: found switch: BCM53125, rev 4
[    1.919994] libphy: bcma_mdio mii bus: probed
[    1.924444] bgmac_bcma bcma0:2: Support for Roboswitch not implemented
[    2.028246] bgmac_bcma: Broadcom 47xx GBit MAC driver loaded
[    2.034421] bcm47xx-wdt bcm47xx-wdt.0: BCM47xx Watchdog Timer enabled (30 seconds)
[    2.042493] PCI: Enabling device 0000:00:01.0 (0000 -> 0002)
[    2.048341] bcma: PCI: Fixing up device 0000:00:01.0
[    2.053407] bcma: change PCIe max read request size from 512 to 128
[    2.059891] bcma-pci-bridge 0000:00:01.0: bus1: Found chip with id 0x4360, rev 0x02 and package 0x01
[    2.069286] bcma-pci-bridge 0000:00:01.0: bus1: Core 0 found: ChipCommon (manuf 0x4BF, id 0x800, rev 0x2A, class 0x0)
[    2.080156] bcma-pci-bridge 0000:00:01.0: bus1: Core 1 found: IEEE 802.11 (manuf 0x4BF, id 0x812, rev 0x28, class 0x0)
[    2.091119] bcma-pci-bridge 0000:00:01.0: bus1: Core 2 found: ARM CR4 (manuf 0x4BF, id 0x83E, rev 0x00, class 0x0)
[    2.101730] bcma-pci-bridge 0000:00:01.0: bus1: Core 3 found: PCIe Gen2 (manuf 0x4BF, id 0x83C, rev 0x00, class 0x0)
[    2.112517] bcma-pci-bridge 0000:00:01.0: bus1: Core 4 found: USB 2.0 Device (manuf 0x4BF, id 0x81A, rev 0x0E, class 0x0)
[    2.147685] bcma-pci-bridge 0000:00:01.0: bus1: Invalid SPROM read from the PCIe card, trying to use fallback SPROM
[    2.207566] bcma-pci-bridge 0000:00:01.0: bus1: Bus registered
[    2.215271] NET: Registered protocol family 10
[    2.227784] Segment Routing with IPv6
[    2.231667] NET: Registered protocol family 17
[    2.236269] 8021q: 802.1Q VLAN Support v1.8
[    2.248867] VFS: Mounted root (squashfs filesystem) readonly on device 31:4.
[    2.256725] Freeing unused kernel memory: 184K
[    2.261336] This architecture does not have kernel memory protection.
[    2.547068] random: fast init done
[    3.266309] init: Console is alive
[    3.270298] init: - watchdog -
[    4.961621] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    5.171656] usbcore: registered new interface driver usbfs
[    5.177494] usbcore: registered new interface driver hub
[    5.183076] usbcore: registered new device driver usb
[    5.197526] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    5.216583] SCSI subsystem initialized
[    5.227783] ehci-platform: EHCI generic platform driver
[    5.233278] ehci-platform ehci-platform.0: EHCI Host Controller
[    5.239457] ehci-platform ehci-platform.0: new USB bus registered, assigned bus number 1
[    5.247899] ehci-platform ehci-platform.0: irq 5, io mem 0x18004000
[    5.277113] ehci-platform ehci-platform.0: USB 2.0 started, EHCI 1.00
[    5.285133] hub 1-0:1.0: USB hub found
[    5.289761] hub 1-0:1.0: 2 ports detected
[    5.301532] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    5.309794] ohci-platform: OHCI generic platform driver
[    5.315250] ohci-platform ohci-platform.0: Generic Platform OHCI controller
[    5.322468] ohci-platform ohci-platform.0: new USB bus registered, assigned bus number 2
[    5.330877] ohci-platform ohci-platform.0: irq 5, io mem 0x18009000
[    5.412519] hub 2-0:1.0: USB hub found
[    5.416981] hub 2-0:1.0: 2 ports detected
[    5.425372] uhci_hcd: USB Universal Host Controller Interface driver
[    5.450731] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    5.468837] init: - preinit -
[    7.756241] random: jshn: uninitialized urandom read (4 bytes read)
[    7.832920] random: jshn: uninitialized urandom read (4 bytes read)
[    7.884797] random: jshn: uninitialized urandom read (4 bytes read)
[    7.934443] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    8.967123] bgmac_bcma bcma0:2 eth0: Link is Up - 1Gbps/Full - flow control off
[    8.974654] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   11.509997] jffs2: notice: (454) jffs2_build_xattr_subsystem: complete building xattr subsystem, 3 of xdatum (0 unchecked, 3 orphan) and 3 of xref (3 dead, 0 orphan) found.
[   11.527837] mount_root: switching to jffs2 overlay
[   11.577280] overlayfs: upper fs does not support tmpfile.
[   11.591135] urandom-seed: Seeding with /etc/urandom.seed
[   11.771357] procd: - early -
[   11.774512] procd: - watchdog -
[   12.591773] procd: - watchdog -
[   12.595699] procd: - ubus -
[   12.691206] random: ubusd: uninitialized urandom read (4 bytes read)
[   12.701753] random: ubusd: uninitialized urandom read (4 bytes read)
[   12.709132] random: ubusd: uninitialized urandom read (4 bytes read)
[   12.717857] procd: - init -
Please press Enter to activate this console.
[   13.795880] kmodloader: loading kernel modules from /etc/modules.d/*
[   13.835227] ip6_tables: (C) 2000-2006 Netfilter Core Team
[   13.861849] Loading modules backported from Linux version v4.19.23-0-g67d52fae61c1
[   13.869633] Backport generated by backports.git v4.19.23-1-0-g480a925a
[   13.888584] ip_tables: (C) 2000-2006 Netfilter Core Team
[   13.922455] nf_conntrack version 0.5.0 (2048 buckets, 8192 max)
[   14.045285] xt_time: kernel timezone is -0000
[   14.140761] PPP generic driver version 2.4.2
[   14.149714] NET: Registered protocol family 24
[   14.172721] b43-phy0: Broadcom 4716 WLAN found (core revision 17)
[   14.179885] b43-phy0: Found PHY: Analog 8, Type 4 (N), Revision 5
[   14.186103] b43-phy0: Found Radio: Manuf 0x17F, ID 0x2056, Revision 7, Version 0
[   14.194647] b43-phy1: Broadcom 4360 WLAN found (core revision 40)
[   14.217266] b43-phy1 ERROR: FOUND UNSUPPORTED PHY (Analog 12, Type 11 (AC), Revision 0)
[   14.225477] b43: probe of bcma1:1 failed with error -122
[   14.231162] Broadcom 43xx driver loaded [ Features: NL ]
[   14.269271] kmodloader: done loading kernel modules from /etc/modules.d/*
[   28.613343] br-lan: port 1(eth0) entered blocking state
[   28.618764] br-lan: port 1(eth0) entered disabled state
[   28.624590] device eth0 entered promiscuous mode
[   28.676554] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   29.687199] bgmac_bcma bcma0:2 eth0: Link is Up - 1Gbps/Full - flow control off
[   29.694751] br-lan: port 1(eth0) entered blocking state
[   29.700158] br-lan: port 1(eth0) entered forwarding state
[   29.747226] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
```

motolav commented:
@hauke M

adrianschmutzler commented:
Hi, please add properly formatted specifications and flashing instructions to your commit message (just have a look at what other recent device support commits did). In principle, you'd just need to add what you posted as first comment in this PR.

Despite, I don't think that patches for kernel 4.14 are still necessary. Please add 5.4 patches instead, unless all of them are actually merged there already.

However, note that I won't merge this, so you will have to rely on hauke or others when this is addressed.

adrianschmutzler commented:
Please add a Signed-off-by to your commit message.

adrianschmutzler commented:
The rest seems fine despite some minor issues, I'd merge this if you add the Signed-off-by.

adrianschmutzler commented on target/linux/bcm47xx/base-files/etc/board.d/01_network:
```diff
@@ -148,6 +148,11 @@ configure_by_model() {
 			"0:lan:4" "1:lan:3" "2:lan:2" "3:lan:1" "4:wan" "5@eth0"
 		;;
 
+	"Netgear R6200 V1")
```
This should be moved to the correct position (alphabetic sorting).

adrianschmutzler commented on target/linux/bcm47xx/patches-4.19/032-v5.4-MIPS-BCM47XX-Add-support-for-Netgear-R6200v1.patch:
```diff
@@ -0,0 +1,46 @@
++	BCM47XX_BOARD_NETGEAR_R6200_V1,
+ 	BCM47XX_BOARD_NETGEAR_WGR614V8,
+ 	BCM47XX_BOARD_NETGEAR_WGR614V9,
+ 	BCM47XX_BOARD_NETGEAR_WGR614_V10,
```
Newline should be added here.

adrianschmutzler commented:
"add" in commit title should be lower-case.

motolav committed:
bcm47xx: add support for NETGEAR R6200 V1

This patch adds supports for the NETGEAR R6200 V1
This device is mainlined in Linux as of 5.4

Specification:
   - SOC: Broadcom BCM4718A1 (480 MHz)
   - Flash: 16MB (Macronix MX25L128)
   - RAM: 128MB DDR2
   - SWITCH: BCM53125
   - Ethernet: 5x GE (1 WAN + 4 LAN, Inverted order)
   - WLAN 2.4G: In SOC
   - WLAN 5GHz: miniPCIe card, BCM4352 (ID: 4360)
   - USB: 1x USB 2.0 port (Type A)
   - Buttons: Reset, WLAN, WPS
   - LED: Amber for Power, others need a workaround
   - UART: 1x UART on PCB

LED Issues: They are controlled by a 74HC164 via bit banging(GPIO 6/7)

Firmware Install Instructions
=============================
Using the device specific .chk, use that file to perform
a normal device upgrade in the OEM admin page.

Signed-off-by: Edward Matijevic <motolav@gmail.com>

motolav commented:
@adrianschmutzler I fixed the issues and thanks

adrianschmutzler commented:
Added to my staging tree.

Thanks for your endurance.
